unison-cli/src/Unison/Codebase/Editor/HandleInput/Update.hs

```diff
@@ -189,12 +201,324 @@ handleUpdate input optionalPatch requestedNames = do
+            nameToInterimInfo =
+              UF.hashTermsId (Slurp.originalFile slurp0)
+
+        -- Associate each term name being updated with its interim reference.
+        --
+        -- Running example:
+        --
+        --   "ping" <=> #newping
+        let nameToInterimRef :: Bij Symbol TermReferenceId
+            nameToInterimRef =
+              nameToInterimInfo
+                & Map.map (\(ref, _wk, _term, _typ) -> ref)
+                & bijFromMap
+
+        -- Remap the references contained in each implicit term, then add back in the slurped interim terms and unhash
+        -- the collection. The unhashing process will invent a fresh variable name for each term.
+        --
+        -- Running example:
+        --
+        --   #newping       => ("fresh1", <"fresh3" + 4>)
+        --   #pingpong.pong => ("fresh2", <"fresh1" + 2>)
+        --   #wham          => ("fresh3", <"fresh2" + 3>)
+        let refToGeneratedNameAndTerm :: Map TermReferenceId (Symbol, Term Symbol Ann)
+            refToGeneratedNameAndTerm =
+              -- Running example:
+              --
+              --   #pingpong.pong => (<#pingpong.ping + 2>, { "pong" })
+              --   #wham          => (<#pingpong.pong + 3>, { "wham" })
+              implicitTerms
+                -- Running example:
+                --
+                --   #pingpong.pong => <#newping + 2>
+                --   #wham          => <#pingpong.pong + 3>
+                & Map.map (\(term, _names) -> rewrite term)
+                -- Running example:
+                --
+                --   #newping       => <#wham + 4>
+                --   #pingpong.pong => <#newping + 2>
+                --   #wham          => <#pingpong.pong + 3>
+                & Map.union interimRefToTerm
+                & Term.unhashComponent
+              where
+                -- Running example:
+                --
+                --   #newping => <#wham + 4>
+                interimRefToTerm :: Map TermReferenceId (Term Symbol Ann)
+                interimRefToTerm =
+                  Map.remap (\(_var, (ref, _wk, term, _typ)) -> (ref, term)) nameToInterimInfo
+                -- Running example: apply the following reference mapping everwhere in a term:
+                --
+                --   #pingpong.ping -> #newping
+                --   ref            -> ref
+                rewrite :: Term Symbol Ann -> Term Symbol Ann
+                rewrite =
+                  updatedNameToOldRefs
+                    & Map.toList
+                    & foldMap
+                      ( \(name, oldRefs) ->
+                          oldRefs
+                            & Set.toList
+                            & map
+                              ( \oldRef ->
+                                  case bijLookupL name nameToInterimRef of
+                                    Nothing -> undefined
```

Comment:
`undefined`; Can this be replaced by some CallStack'd error call or something similar please?

Reply:
Yeah- whoops! Meant to add a better error message.